File: src/main/java/no/fintlabs/mapping/InstanceFlowHeadersMappingService.java
package no.fintlabs.mapping;

import no.fintlabs.flyt.kafka.headers.InstanceFlowHeaders;
import no.fintlabs.repository.entities.InstanceFlowHeadersEmbeddable;
import org.springframework.stereotype.Service;

@Service
public class InstanceFlowHeadersMappingService {

    public InstanceFlowHeadersEmbeddable toEmbeddable(InstanceFlowHeaders instanceFlowHeaders) {
        return InstanceFlowHeadersEmbeddable.builder()
                .sourceApplicationId(instanceFlowHeaders.getSourceApplicationId())
                .sourceApplicationIntegrationId(instanceFlowHeaders.getSourceApplicationIntegrationId())
                .sourceApplicationInstanceId(instanceFlowHeaders.getSourceApplicationInstanceId())
                .fileIds(instanceFlowHeaders.getFileIds())
                .correlationId(instanceFlowHeaders.getCorrelationId())
                .integrationId(instanceFlowHeaders.getIntegrationId())
                .instanceId(instanceFlowHeaders.getInstanceId())
                .configurationId(instanceFlowHeaders.getConfigurationId())
                .archiveInstanceId(instanceFlowHeaders.getArchiveInstanceId())
                .build();
    }

    public InstanceFlowHeaders toInstanceFlowHeaders(InstanceFlowHeadersEmbeddable embeddable) {
        return InstanceFlowHeaders.builder()
                .sourceApplicationId(embeddable.getSourceApplicationId())
                .sourceApplicationIntegrationId(embeddable.getSourceApplicationIntegrationId())
                .sourceApplicationInstanceId(embeddable.getSourceApplicationInstanceId())
                .fileIds(embeddable.getFileIds())
                .correlationId(embeddable.getCorrelationId())
                .integrationId(embeddable.getIntegrationId())
                .instanceId(embeddable.getInstanceId())
                .configurationId(embeddable.getConfigurationId())
                .archiveInstanceId(embeddable.getArchiveInstanceId())
                .build();
    }

}

File: src/main/kotlin/no/novari/flyt/history/mapping/InstanceFlowHeadersMappingService.kt
package no.novari.flyt.history.mapping

import no.novari.flyt.history.repository.entities.InstanceFlowHeadersEmbeddable
import no.novari.flyt.kafka.instanceflow.headers.InstanceFlowHeaders
import org.springframework.stereotype.Service

@Service
class InstanceFlowHeadersMappingService {
    fun toEmbeddable(instanceFlowHeaders: InstanceFlowHeaders): InstanceFlowHeadersEmbeddable {
        return InstanceFlowHeadersEmbeddable(
            sourceApplicationId = instanceFlowHeaders.sourceApplicationId,
            sourceApplicationIntegrationId = instanceFlowHeaders.sourceApplicationIntegrationId,
            sourceApplicationInstanceId = instanceFlowHeaders.sourceApplicationInstanceId,
            fileIds = instanceFlowHeaders.fileIds?.toMutableList() ?: mutableListOf(),
            correlationId = instanceFlowHeaders.correlationId,
            integrationId = instanceFlowHeaders.integrationId,
            instanceId = instanceFlowHeaders.instanceId,
            configurationId = instanceFlowHeaders.configurationId,
            archiveInstanceId = instanceFlowHeaders.archiveInstanceId,
        )
    }

    fun toInstanceFlowHeaders(embeddable: InstanceFlowHeadersEmbeddable): InstanceFlowHeaders {
        val builder =
            InstanceFlowHeaders
                .builder()
                .sourceApplicationId(requireNotNull(embeddable.sourceApplicationId))
                .sourceApplicationIntegrationId(requireNotNull(embeddable.sourceApplicationIntegrationId))
                .sourceApplicationInstanceId(requireNotNull(embeddable.sourceApplicationInstanceId))
                .fileIds(embeddable.fileIds)
                .integrationId(embeddable.integrationId)
                .instanceId(embeddable.instanceId)
                .configurationId(embeddable.configurationId)
                .archiveInstanceId(embeddable.archiveInstanceId)

        embeddable.correlationId?.let(builder::correlationId)

        return builder.build()
    }
}
